### ISSUE TITLE
[ENH] `gluonts` models

### ISSUE THREAD

**Issue by fkiraly** [sktime/sktime#6800]
It would be nice to have `gluonts` models interface and indexed.

These are mainly a handful of torch/lightning based models with a consistent, `torch` based interface.

There is recent work towards adding the `gluonts` data containers as mtypes, #6530, which should allow to add these easily.

Contingent on 6530 being merged - good first issue in collaboration with current workstreams.

**Comment by fkiraly** [comment]
@rajatkriplani, yes, that should work, with some small comments:

* there are already existing data types for `gluonts` internal data types, they are not used yet though in too many places.
* a common adapter class with different models being different instances or descendants might avoid code duplication, although writing two or there of them and then refactoring might be easier than planning it from scratch.

**Comment by rajatkriplani** [comment]
@fkiraly Great! I will start working on it, keep you updated and ask for feedback/questions on discord.

**Comment by fkiraly** [comment]
Excellent!

FYI, I have also wrapped up the API documentation on in-memory data types.

This should help with finding out how to use the `gluonts` data types, although some documentation on those is missing.
The respective issue is here: https://github.com/sktime/sktime/issues/7386

**Comment by rajatkriplani** [comment]
@fkiraly below is the summary based on our discussion and your feedback:
here’s how a more informal, human-like summary would look:

1. **initial discussion**:  
   the idea is to integrate gluonts models into `sktime` using the gluonts data containers added in #6530.

2. **proposed implementation**:  
   we’re using a modular approach with a `GluonTSBaseAdapter` and subclasses like `DeepARAdapter` and `NBEATSAdapter`.

3. **feedback and refinements**:  
   added an `_init_gluonts_model` method to handle model setup, which makes it less repetitive. also moved `Trainer` initialization into `_init_gluonts_model`.

4. **unit tests**:  
   wrote tests for data conversion, `fit`/`predict` methods, and ensuring `_init_gluonts_model` raises an error if not implemented.

5. **current status**:  
   so far, `GluonTSBaseAdapter`, `DeepARAdapter`, and `NBEATSAdapter` are implemented and tested with sample data.

6. **next steps**:  
   should i add more models now, or is the current progress good enough for the time being and create a PR?

**Comment by rajatkriplani** [comment]
@fkiraly 

1. **Initial discussion**:  
   Idea was to integrate gluonts models into `sktime` using the gluonts data containers added in #6530.

2. **Proposed Implementation**:  
   Using a modular approach with a `GluonTSBaseAdapter` and subclasses like `DeepARAdapter` and `NBEATSAdapter`.

3. **feedback and refinements**:  
   Added an `_init_gluonts_model` method to handle model setup, which makes it less repetitive. also moved `Trainer` initialization into `_init_gluonts_model`.

4. **Unit tests**:  
   Wrote tests for data conversion, `fit`/`predict` methods, and ensuring `_init_gluonts_model` raises an error if not implemented.

5. **Current status**:  
   so far, `GluonTSBaseAdapter`, `DeepARAdapter`, and `NBEATSAdapter` are implemented and tested with sample data.

6. Next steps:  
   Is the current progress good enough for the time being?

**Comment by fkiraly** [comment]
Yes, absolutely! It would be great if you could paste your code draft from discord as well, so other developers can also comment, and things are in a single place.

**Comment by rajatkriplani** [comment]
Hi @fkiraly I was reviewing my code to create PR but then I forgot something. Earlier I mentioned in of of my previous comment:
> Use new GluonTS data conversions in `sktime/datatypes/_adapter/gluonts.py` to handle data formatting.

and I didn't do it, now I have modified `gluonts_adapters.py` please review it:
```python
import pandas as pd
from sktime.forecasting.base import BaseForecaster
from gluonts.dataset.common import ListDataset
from gluonts.mx.trainer import Trainer

from sktime.datatypes._adapter.gluonts import (
    convert_gluonts_result_to_multiindex,
    convert_from_multiindex_to_listdataset,
)


class GluonTSBaseAdapter(BaseForecaster):
    """
    Common adapter base class for GluonTS models.
    """

    def __init__(self, model_class, freq="D", prediction_length=24, epochs=10, **kwargs):
        """
        Parameters
        ----------
        model_class : class
            GluonTS model class to instantiate.
        freq : str, optional (default="D")
            Frequency of the time series.
        prediction_length : int, optional (default=24)
            Number of time steps to predict.
        epochs : int, optional (default=10)
            Number of training epochs.
        kwargs : dict
            Additional keyword arguments for the GluonTS model.
        """
        self.freq = freq
        self.prediction_length = prediction_length
        self.epochs = epochs
        self.model_class = model_class
        self.model_kwargs = kwargs

        self.model = self._init_model()
        super().__init__()

    def _init_model(self):
        """
        Initialize the specific GluonTS model with a generic trainer.
        """
        trainer = Trainer(epochs=self.epochs)
        return self.model_class(
            freq=self.freq,
            prediction_length=self.prediction_length,
            trainer=trainer,
            **self.model_kwargs,
        )

    def _convert_to_gluonts_format(self, y):
        """
        Convert sktime data to GluonTS ListDataset format.
        """
        return convert_from_multiindex_to_listdataset(y)

    def _convert_to_sktime_format(self, gluonts_result):
        """
        Convert GluonTS forecast result to sktime format.
        """
        return convert_gluonts_result_to_multiindex(gluonts_result)

    def _fit(self, y, X=None, fh=None):
        """
        Fit the model to the training data.
        """
        gluonts_data = self._convert_to_gluonts_format(y)
        self.model.train(gluonts_data)
        return self

    def _predict(self, fh, X=None):
        """
        Generate predictions.
        """
        gluonts_data = self._convert_to_gluonts_format(self._y)
        forecast_it, ts_it = self.model.make_evaluation_predictions(
            gluonts_data, num_samples=100
        )
        forecast_entry = next(forecast_it)
        return self._convert_to_sktime_format(forecast_entry)

    def update(self, y_new, update_params=True):
        """
        Update or retrain the model with new data.
        """
        gluonts_data = self._convert_to_gluonts_format(y_new)
        self.model.train(gluonts_data)
        return self


class DeepARAdapter(GluonTSBaseAdapter):
    """
    Adapter for the GluonTS DeepAR model.
    """

    def __init__(self, **kwargs):
        from gluonts.model.deepar import DeepAREstimator

        super().__init__(model_class=DeepAREstimator, **kwargs)


class NBEATSAdapter(GluonTSBaseAdapter):
    """
    Adapter for the GluonTS N-BEATS model.
    """

    def __init__(self, **kwargs):
        from gluonts.model.n_beats import NBEATSEstimator

        super().__init__(model_class=NBEATSEstimator, **kwargs)

```

**Comment by fkiraly** [comment]
@rajatkriplani, great! Code like this is better to review as a pull request - would you kindly open one?

**Comment by rajatkriplani** [comment]
Hi @fkiraly Can help me get all the checks on my PR?
![Screenshot Capture - 2024-12-31 - 19-34-35](https://github.com/user-attachments/assets/b58d8e58-3b69-4ef8-a5d9-ec0dbe7369e1)

**Comment by fkiraly** [comment]
(I believe @yarnabrina already commented there)
